test/test_functional.ml

```diff
@@ -0,0 +1,222 @@
+  assert_equal_string ~ctxt expected got
+
+let test_verify (module R : P11.S) session key =
+  let verify signature =
+    let mechanism = P11.Mechanism.CKM_SHA256_RSA_PKCS in
+    let data = "some data" in
+    R.verify session mechanism key ~data ~signature
+  in
+  let test_valid ctxt =
+    assert_equal
+      ~ctxt
+      ()
+      (verify "OK")
+  in
+  let test_invalid ctxt =
+    assert_raises (P11.CKR P11.RV.CKR_SIGNATURE_INVALID) @@ fun () ->
+    verify "Invalid"
+  in
+  [ "Valid" >:: test_valid;
+    "Invalid" >:: test_invalid;
+  ]
+
+let test_wrap (module R : P11.S) session key ctxt =
+  let mechanism = P11.Mechanism.CKM_AES_ECB in
+  let expected = "fake wrapped" in
+  let got = R.wrap_key session mechanism key key in
+  assert_equal_string ~ctxt expected got
+
+let test_unwrap (module R : P11.S) session key ctxt =
+  let mechanism = P11.Mechanism.CKM_AES_ECB in
+  let expected = Unsigned.ULong.of_int 1 in
+  let plaintext = "plaintext" in
+  let got = R.unwrap_key session mechanism key plaintext [] in
+  assert_equal_object_handle ~ctxt expected got
+
+let test_driver driver =
+  [ "Get info" >:: test_get_info driver;
+    "Get slot list" >:: test_get_slot_list driver;
+  ]
+
+let open_session (module R : P11.S) =
+  let session =
+    R.open_session
+      ~slot:Unsigned.ULong.zero
+      ~flags:P11.Flags._CKF_SERIAL_SESSION
+  in
+  R.login session P11.User_type.CKU_USER "1234";
+  session
+
+let test_session driver =
+  let session = open_session driver in
+  [ "Create object" >:: test_create_object driver session;
+    "Generate key pair" >:: test_generate_key_pair driver session;
+  ]
+
+let test_object driver =
+  let session = open_session driver in
+  let (module R : P11.S) = driver in
+  let key =
+    R.generate_key
+      session
+      P11.Mechanism.CKM_AES_KEY_GEN
+      [ P11.Attribute.Pack (
+            P11.Attribute_type.CKA_VALUE_LEN,
+            Unsigned.ULong.of_int 32
+          )
+      ]
+  in
+  [ "Encrypt" >:: test_encrypt driver session key;
+    "Decrypt" >:: test_decrypt driver session key;
+    "Copy object" >:: test_copy_object driver session key;
+    "Get attribute value" >::: test_get_attribute_value driver session key;
+    "Set attribute value" >:: test_set_attribute_value driver session key;
+    "Derive" >:: test_derive driver session key;
+    "Sign" >:: test_sign driver session key;
+    "Verify" >::: test_verify driver session key;
+    "Wrap" >:: test_wrap driver session key;
+    "Unwrap" >:: test_unwrap driver session key;
+  ]
+
+let test_p11_load =
+  let dll = "./_build/test/dllpkcs11_fake.so" in
+  let driver =
+    P11.load_driver
+      ?log_calls:None
+      ?on_unknown:None
+      ~use_get_function_list:`False
+      ~dll
+  in
+  let (module R) = driver in
+  R.initialize ();
+  [ "Driver" >::: test_driver driver;
+    "Session" >::: test_session driver;
+    "Object" >::: test_object driver;
+  ]
+
+let suite =
+  "Functional" >::: [
+    "Load P11" >::: test_p11_load;
```

Comment:
There's a mix of comma-first and other styles.  Full comma-first style would be preferable no?